src/Provider/Facebook.php

```diff
@@ -83,6 +90,55 @@ protected function initialize()
         }
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function apiRequest($url, $method = 'GET', $parameters = [], $headers = [], $multipart = false)
+    {
+        // Handle token exchange prior to the standard handler for an API request
+        $exchange_by_expiry_days = $this->config->get('exchange_by_expiry_days') ?? 45;
+        if ($exchange_by_expiry_days !== null) {
+            $projected_timestamp = time() + 60 * 60 * 24 * $exchange_by_expiry_days;
+            if (($this->hasAccessTokenExpired() === false) && // Not expired yet
+                ($this->hasAccessTokenExpired($projected_timestamp) === true) // But will within projection
+            ) {
+                $this->exchangeAccessToken();
+            }
+        }
+
+        return parent::apiRequest($url, $method, $parameters, $headers, $multipart);
+    }
+
+    /**
+     * Exchange the Access Token with one that expires further in the future.
+     *
+     * @return string Raw Provider API response
+     * @throws \Hybridauth\Exception\HttpClientFailureException
+     * @throws \Hybridauth\Exception\HttpRequestFailedException
+     * @throws InvalidAccessTokenException
+     */
+    public function exchangeAccessToken()
+    {
+        $exchangeTokenParameters = [
+            'grant_type'        => 'fb_exchange_token',
+            'client_id'         => $this->clientId,
+            'client_secret'     => $this->clientSecret,
+            'fb_exchange_token' => $this->getStoredData('access_token'),
```

Comment:
let's say we get long-lived token, and then it will be stored as `access_token`? - next call of `exchangeAccessToken` will use long-lived token and this will not work. Since, we need to pass short-lived token.

did I miss something?

Reply:
Regarding your last comment, short-lived and long-lived tokens are both are access tokens, it just varies by expiry. It's not like the 'code' parameter during OAuth flow that gets exchanged to a regular tokens.